plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/Fabric8AuthServiceClient.java

```diff
@@ -0,0 +1,206 @@
+ */
+package com.redhat.che.multitenant;
+
+import com.google.common.io.CharStreams;
+import com.google.gson.FieldNamingPolicy;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.inject.name.Named;
+import io.jsonwebtoken.Jwt;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.ws.rs.HttpMethod;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriBuilder;
+import org.eclipse.che.api.core.BadRequestException;
+import org.eclipse.che.api.core.ForbiddenException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.UnauthorizedException;
+import org.eclipse.che.commons.env.EnvironmentContext;
+import org.eclipse.che.commons.lang.Pair;
+import org.eclipse.che.dto.server.DtoFactory;
+import org.eclipse.che.multiuser.keycloak.server.KeycloakServiceClient;
+import org.eclipse.che.multiuser.keycloak.server.KeycloakSettings;
+import org.eclipse.che.multiuser.keycloak.shared.dto.KeycloakErrorResponse;
+
+/**
+ * Abstraction of fabric8 auth server, allowing for account linking and obtaining github tokens.
+ *
+ * @author Angel Misevski (amisevsk@redhat.com)
+ */
+@Singleton
+public class Fabric8AuthServiceClient extends KeycloakServiceClient {
+
+  private static final String GITHUB_LINK_API_PATH = "/api/token/link?for=https://github.com";
+  private static final String GITHUB_TOKEN_API_PATH = "/api/token?for=https://github.com";
+  private static final Gson gson =
+      new GsonBuilder()
+          .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
+          .create();
+
+  private final String githubTokenEndpoint;
+  private final String githubLinkEndpoint;
+
+  @Inject
+  public Fabric8AuthServiceClient(
+      @Named("che.fabric8.auth.endpoint") String baseAuthUrl, KeycloakSettings keycloakSettings) {
+    super(keycloakSettings);
+    this.githubTokenEndpoint = baseAuthUrl + GITHUB_TOKEN_API_PATH;
+    this.githubLinkEndpoint = baseAuthUrl + GITHUB_LINK_API_PATH;
+  }
+
+  /**
+   * Return account linking url for user. Note that this does not return the fabric8 auth linking
+   * URL, but instead makes a request to that url and returns the redirect obtained from there. This
+   * is because requests to fabric8 auth need the Authorization header, and a user's browser will
+   * not include this header by default, so returning the direct URL does not work.
+   */
+  @Override
+  public String getAccountLinkingURL(
+      @SuppressWarnings("rawtypes") Jwt token, String oauthProvider, String redirectAfterLogin) {
+    String linkingEndpoint =
+        UriBuilder.fromUri(githubLinkEndpoint)
```

Comment:
Don't you think it would be better to get the `tokenEndpoint` and `linkEndpoint` from `fabric8.*` properties whose name would derive from the `oauthProvider` name ? That would make it much more generic instead of having hard-wired for the GitHub provider ?

Reply:
@davidfestal do you mean adding extra properties to cm for:

- https://auth.openshift.io/api/token/link?for=openshift 
- https://auth.openshift.io/api/token?for=openshift

Reply:
yes, more precisely, I mean something like:
```
che.fabric8.oauth.provider.github.endpoints.link=api/token/link?for=https://github.com
che.fabric8.oauth.provider.github.endpoints.token=api/token?for=https://github.com
```
And of course it would be available for Openshift as well.

Reply:
see this code: https://github.com/redhat-developer/rh-che/blob/master/plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/OpenshiftUserTokenProvider.java#L74
seems that this property is in fact already available as
```
 @Nullable @Named(OIDC_PROVIDER_SETTING) String oidcProvider
```
The only difference is that it already contains the `/api` suffix

Reply:
@amisevsk @davidfestal I would propose to merge the PR in the current state and update / cleanup properties in a separate issue / PR. I believe there is smth. we could definitely cleanup according to the - https://github.com/redhat-developer/rh-che/pull/810#issuecomment-415852628